cpp/log/cluster_state_controller-inl.h

```diff
@@ -0,0 +1,203 @@
+          std::bind(&ClusterStateController<Logged>::ClusterServingSTHUpdater,
+                    this)) {
+  store_->WatchClusterNodeStates(
+      std::bind(&ClusterStateController::OnClusterStateUpdated, this,
+                std::placeholders::_1));
+}
+
+
+template <class Logged>
+ClusterStateController<Logged>::~ClusterStateController() {
+  {
+    std::unique_lock<std::mutex> lock(mutex_);
+    exiting_ = true;
+    update_required_cv_.notify_all();
+  }
+  cluster_serving_sth_update_thread_.join();
+}
+
+
+template <class Logged>
+void ClusterStateController<Logged>::NewTreeHead(
+    const ct::SignedTreeHead& sth) {
+  CHECK(sth.IsInitialized());
+  std::unique_lock<std::mutex> lock(mutex_);
+  if (local_node_state_.has_newest_sth()) {
+    CHECK_GE(sth.timestamp(), local_node_state_.newest_sth().timestamp());
+  }
+  local_node_state_.mutable_newest_sth()->CopyFrom(sth);
+  PushLocalNodeState(lock);
+}
+
+
+template <class Logged>
+void ClusterStateController<Logged>::ContiguousTreeSizeUpdated(
+    const int64_t new_contiguous_tree_size) {
+  CHECK_GE(new_contiguous_tree_size, 0);
+  std::unique_lock<std::mutex> lock(mutex_);
+  CHECK_GE(new_contiguous_tree_size, local_node_state_.contiguous_tree_size());
+  local_node_state_.set_contiguous_tree_size(new_contiguous_tree_size);
+  PushLocalNodeState(lock);
+}
+
+
+template <class Logged>
+bool ClusterStateController<Logged>::GetCalculatedServingSTH(
+    SignedTreeHead* sth) const {
+  CHECK_NOTNULL(sth);
+  std::lock_guard<std::mutex> lock(mutex_);
+  if (calculated_serving_sth_.get() == nullptr) {
+    return false;
+  }
+  *sth = *calculated_serving_sth_;
+  return true;
+}
+
+
+template <class Logged>
+void ClusterStateController<Logged>::PushLocalNodeState(
+    const std::unique_lock<std::mutex>& lock) {
+  CHECK(lock.owns_lock());
+  const util::Status status(store_->SetClusterNodeState(local_node_state_));
+  if (!status.ok()) {
+    LOG(WARNING) << status;
+  }
+}
+
+
+template <class Logged>
+void ClusterStateController<Logged>::OnClusterStateUpdated(
+    const std::vector<Update<ct::ClusterNodeState>>& updates) {
+  std::unique_lock<std::mutex> lock(mutex_);
+  for (const auto& update : updates) {
+    const std::string& node_id(update.handle_.Entry().node_id());
+    if (update.exists_) {
+      all_node_states_[node_id] = update.handle_.Entry();
+    } else {
+      CHECK_EQ(1, all_node_states_.erase(node_id));
+    }
+  }
+
+  if (CalculateServingSTH(lock) && election_->IsMaster()) {
+    update_required_ = true;
+    update_required_cv_.notify_all();
+  }
+}
+
+
+template <class Logged>
+bool ClusterStateController<Logged>::CalculateServingSTH(
+    const std::unique_lock<std::mutex>& lock) {
+  VLOG(1) << "Calculating new ServingSTH...";
+  CHECK(lock.owns_lock());
+
+  // First, create a mapping of tree size to number of nodes at that size, and
+  // a mapping of the newst STH for any given size:
+  std::map<uint64_t, ct::SignedTreeHead> sth_by_size;
```

Comment:
I think the ```uint64_t``` should be ```int64_t``` instead? It always feels weird to me as well to use signed for sizes, but there are some decent arguments for it...